extmod/vfs_blockdev.c

```diff
@@ -48,8 +77,10 @@ void mp_vfs_blockdev_init(mp_vfs_blockdev_t *self, mp_obj_t bdev) {
 
 int mp_vfs_blockdev_read(mp_vfs_blockdev_t *self, size_t block_num, size_t num_blocks, uint8_t *buf) {
     if (self->flags & MP_BLOCKDEV_FLAG_NATIVE) {
-        mp_uint_t (*f)(uint8_t *, uint32_t, uint32_t) = (void *)(uintptr_t)self->readblocks[2];
-        return f(buf, block_num, num_blocks);
+        // CIRCUITPY-CHANGE: Pass the blockdev object into native readblocks so
+        // it has the corresponding state.
+        mp_uint_t (*f)(mp_obj_t self, uint8_t *, uint32_t, uint32_t) = (void *)(uintptr_t)self->readblocks[2];
```

Comment:
This cast would be an undetected error if we had any other block device implementations (micropython has several) but we control them all it's probably OK.